File: server/engine/business_rule_heuristics.py
"""Direct port of src/lib/generators/businessRuleHeuristics.ts."""

from __future__ import annotations

import re
from dataclasses import dataclass

TRANSFORMATION_STRATEGIES = (
    "DIRECT_COPY",
    "CASE_EXPRESSION",
    "DEFAULT_OR_LOOKUP",
    "CONCAT_EXPRESSION",
    "DIRECT_SQL_FUNCTION",
    "ARITHMETIC_EXPRESSION",
    "MANUAL_REVIEW",
)


@dataclass
class TransformationClassification:
    strategy: str
    expression: str | None
    raw_text: str


_TRIVIAL_SENTINELS = {
    "same as source",
    "direct map",
    "direct mapping",
    "1:1",
    "no transformation",
    "no transformation required",
    "copy",
    "as-is",
    "as is",
    "direct copy",
    "none",
    "-",
    "n/a",
}

_SQL_KEYWORDS = {
    "case", "when", "then", "else", "end", "and", "or", "not", "null", "as", "is", "in", "like", "between",
    "date", "string", "int", "integer", "bigint", "smallint", "tinyint", "decimal", "numeric", "double",
    "float", "real", "timestamp", "boolean", "varchar", "char", "long", "short", "byte", "binary", "array",
    "map", "struct",
}

_WHITELISTED_FUNCTIONS = {
    "substr", "substring", "trim", "ltrim", "rtrim", "cast", "upper", "lower", "round", "replace",
    "concat", "coalesce", "nvl", "ifnull", "to_date", "date_format", "length", "abs",
}

_AGGREGATE_FUNCTIONS = {"sum", "avg", "count", "min", "max"}


def _is_balanced_parens(text: str) -> bool:
    depth = 0
    for ch in text:
        if ch == "(":
            depth += 1
        if ch == ")":
            depth -= 1
        if depth < 0:
            return False
    return depth == 0


def _strip_string_literals(text: str) -> str:
    text = re.sub(r"'[^']*'", "''", text)
    text = re.sub(r'"[^"]*"', '""', text)
    return text


def _contains_only_whitelisted_tokens(text: str, known_fields: list[str], allow_aggregates: bool) -> bool:
    known_field_set = {f.lower() for f in known_fields}
    without_literals = _strip_string_literals(text)
    tokens = re.findall(r"[A-Za-z_][A-Za-z0-9_]*", without_literals)
    for tok in tokens:
        lower = tok.lower()
        if (
            lower in known_field_set
            or lower in _SQL_KEYWORDS
            or lower in _WHITELISTED_FUNCTIONS
            or (allow_aggregates and lower in _AGGREGATE_FUNCTIONS)
        ):
            continue
        return False
    return True


def is_safe_sql_expression(text: str, known_fields: list[str], allow_aggregates: bool = False) -> bool:
    return _is_balanced_parens(text) and _contains_only_whitelisted_tokens(text, known_fields, allow_aggregates)


def _is_trivial(text: str) -> bool:
    return text.strip().lower() in _TRIVIAL_SENTINELS


def _try_case_expression(text: str) -> str | None:
    if re.search(r"\bcase\b[\s\S]*\bwhen\b[\s\S]*\bend\b", text, re.IGNORECASE):
        return text.strip()
    match = re.match(r"^if\s+(.+?)\s+then\s+(.+?)(?:\s+else\s+(.+?))?$", text.strip(), re.IGNORECASE)
    if match:
        condition, then_val, else_val = match.groups()
        else_clause = f" ELSE {else_val.strip()}" if else_val else ""
        return f"CASE WHEN {condition.strip()} THEN {then_val.strip()}{else_clause} END"
    return None


def _try_default_or_lookup(text: str) -> str | None:
    if re.search(r"\bcoalesce\s*\(", text, re.IGNORECASE):
        return text.strip()
    match = re.match(r"^(.+?)\s+defaults?\s+to\s+(.+)$", text, re.IGNORECASE)
    if match:
        source_expr, default_val = match.groups()
        return f"COALESCE({source_expr.strip()}, {default_val.strip()})"
    return None


def _try_concat_expression(text: str) -> str | None:
    if re.search(r"\bconcat\s*\(", text, re.IGNORECASE):
        return text.strip()
    if "||" in text:
        parts = [p.strip() for p in text.split("||") if p.strip()]
        if len(parts) >= 2:
            return f"CONCAT({', '.join(parts)})"
    if "+" in text and re.search(r"'[^']*'", text):
        parts = [p.strip() for p in text.split("+") if p.strip()]
        if len(parts) >= 2:
            return f"CONCAT({', '.join(parts)})"
    return None


def _try_direct_sql_function(text: str, allow_aggregates: bool) -> str | None:
    functions = _WHITELISTED_FUNCTIONS | _AGGREGATE_FUNCTIONS if allow_aggregates else _WHITELISTED_FUNCTIONS
    pattern = r"\b(" + "|".join(functions) + r")\s*\("
    if re.search(pattern, text, re.IGNORECASE):
        return text.strip()
    return None


def _try_arithmetic_expression(text: str) -> str | None:
    trimmed = text.strip()
    if re.match(r"^[A-Za-z0-9_.\s+\-*/%()]+$", trimmed) and re.search(r"[+\-*/%]", trimmed):
        return trimmed
    return None


def classify_transformation(
    raw_text: str, known_fields: list[str], allow_aggregates: bool = False
) -> TransformationClassification:
    text = raw_text.strip()

    if not text or _is_trivial(text):
        return TransformationClassification(strategy="DIRECT_COPY", expression=None, raw_text=raw_text)

    attempts = [
        ("CASE_EXPRESSION", _try_case_expression(text)),
        ("DEFAULT_OR_LOOKUP", _try_default_or_lookup(text)),
        ("CONCAT_EXPRESSION", _try_concat_expression(text)),
        ("DIRECT_SQL_FUNCTION", _try_direct_sql_function(text, allow_aggregates)),
        ("ARITHMETIC_EXPRESSION", _try_arithmetic_expression(text)),
    ]

    for strategy, expression in attempts:
        if expression is None:
            continue
        if not _is_balanced_parens(expression):
            continue
        if not _contains_only_whitelisted_tokens(expression, known_fields, allow_aggregates):
            continue
        return TransformationClassification(strategy=strategy, expression=expression, raw_text=raw_text)

    return TransformationClassification(strategy="MANUAL_REVIEW", expression=None, raw_text=raw_text)


def qualify_field_references(expression: str, known_fields: list[str], alias: str) -> str:
    result = expression
    sorted_fields = sorted((f for f in known_fields if f), key=len, reverse=True)
    for field in sorted_fields:
        escaped = re.escape(field)
        pattern = re.compile(rf"(?<!['\"`\w]){escaped}(?!['\"`\w])", re.IGNORECASE)
        # A lambda replacement (rather than a plain replacement string) keeps this a literal
        # substitution regardless of what characters happen to be in `field` -- re.sub treats a
        # string replacement's backslashes as backreference escapes, which a literal field name
        # should never be interpreted as.
        result = pattern.sub(lambda _match, f=field: f"{alias}.`{f}`", result)
    return result

File: src/lib/generators/businessRuleHeuristics.ts
export type TransformationStrategy =
  | 'DIRECT_COPY'
  | 'CASE_EXPRESSION'
  | 'DEFAULT_OR_LOOKUP'
  | 'CONCAT_EXPRESSION'
  | 'DIRECT_SQL_FUNCTION'
  | 'ARITHMETIC_EXPRESSION'
  | 'MANUAL_REVIEW';

export interface TransformationClassification {
  strategy: TransformationStrategy;
  /** SQL expression (unqualified field names) ready to be aliased, or null when not parseable. */
  expression: string | null;
  rawText: string;
}

const TRIVIAL_SENTINELS = new Set([
  'same as source',
  'direct map',
  'direct mapping',
  '1:1',
  'no transformation',
  'no transformation required',
  'copy',
  'as-is',
  'as is',
  'direct copy',
  'none',
  '-',
  'n/a',
]);

const SQL_KEYWORDS = new Set([
  'case', 'when', 'then', 'else', 'end', 'and', 'or', 'not', 'null', 'as', 'is', 'in', 'like', 'between',
  // Bare boolean literals -- e.g. "CASE WHEN active_flag = true THEN ... END" -- are SQL keywords,
  // not user-controllable identifiers, exactly like NULL above; omitting them wrongly rejected an
  // otherwise-fully-recognized CASE expression just for containing the word "true"/"false".
  'true', 'false',
  // SQL type names, needed to whitelist CAST(expr AS <TYPE>) expressions
  'date', 'string', 'int', 'integer', 'bigint', 'smallint', 'tinyint', 'decimal', 'numeric', 'double',
  'float', 'real', 'timestamp', 'boolean', 'varchar', 'char', 'long', 'short', 'byte', 'binary', 'array', 'map', 'struct',
]);

const WHITELISTED_FUNCTIONS = new Set([
  'substr', 'substring', 'trim', 'ltrim', 'rtrim', 'cast', 'upper', 'lower', 'round', 'replace',
  'concat', 'coalesce', 'nvl', 'ifnull', 'to_date', 'date_format', 'length', 'abs',
]);

// Only offered to callers that opt in via `allowAggregates` (e.g. the L3 dashboard-KPI generator,
// which computes one summary value across a whole table). Row-level generators must NOT allow
// these -- a bare aggregate has no valid meaning in a per-row correlated comparison and would
// produce broken SQL there.
const AGGREGATE_FUNCTIONS = new Set(['sum', 'avg', 'count', 'min', 'max']);

function isBalancedParens(text: string): boolean {
  let depth = 0;
  for (const ch of text) {
    if (ch === '(') depth++;
    if (ch === ')') depth--;
    if (depth < 0) return false;
  }
  return depth === 0;
}

/** Strips single/double-quoted string literals before token-scanning, so literal contents aren't checked. */
function stripStringLiterals(text: string): string {
  return text.replace(/'[^']*'/g, "''").replace(/"[^"]*"/g, '""');
}

function containsOnlyWhitelistedTokens(text: string, knownFields: string[], allowAggregates: boolean): boolean {
  const knownFieldSet = new Set(knownFields.map((f) => f.toLowerCase()));
  const withoutLiterals = stripStringLiterals(text);
  const tokens = withoutLiterals.match(/[A-Za-z_][A-Za-z0-9_]*/g) ?? [];
  return tokens.every((tok) => {
    const lower = tok.toLowerCase();
    return (
      knownFieldSet.has(lower) ||
      SQL_KEYWORDS.has(lower) ||
      WHITELISTED_FUNCTIONS.has(lower) ||
      (allowAggregates && AGGREGATE_FUNCTIONS.has(lower))
    );
  });
}

/**
 * Same balanced-parens + token-whitelist gate `classifyTransformation` applies to its own
 * regex-matched candidates, exposed standalone so any OTHER source of a candidate SQL expression
 * (e.g. an LLM-assisted parse of a prose transformation rule) can be validated before being
 * trusted -- never take an external suggestion's word for its own safety.
 */
export function isSafeSqlExpression(text: string, knownFields: string[], allowAggregates = false): boolean {
  return isBalancedParens(text) && containsOnlyWhitelistedTokens(text, knownFields, allowAggregates);
}

function isTrivial(text: string): boolean {
  return TRIVIAL_SENTINELS.has(text.trim().toLowerCase());
}

function tryCaseExpression(text: string): string | null {
  if (/\bcase\b[\s\S]*\bwhen\b[\s\S]*\bend\b/i.test(text)) {
    return text.trim();
  }
  const ifThenMatch = text
    .trim()
    .match(/^if\s+(.+?)\s+then\s+(.+?)(?:\s+else\s+(.+?))?$/i);
  if (ifThenMatch) {
    const [, condition, thenVal, elseVal] = ifThenMatch;
    const elseClause = elseVal ? ` ELSE ${elseVal.trim()}` : '';
    return `CASE WHEN ${condition.trim()} THEN ${thenVal.trim()}${elseClause} END`;
  }
  return null;
}

function tryDefaultOrLookup(text: string): string | null {
  if (/\bcoalesce\s*\(/i.test(text)) return text.trim();
  const defaultsToMatch = text.match(/^(.+?)\s+defaults?\s+to\s+(.+)$/i);
  if (defaultsToMatch) {
    const [, sourceExpr, defaultVal] = defaultsToMatch;
    return `COALESCE(${sourceExpr.trim()}, ${defaultVal.trim()})`;
  }
  if (/\blookup\b/i.test(text)) return null;
  return null;
}

function tryConcatExpression(text: string): string | null {
  if (/\bconcat\s*\(/i.test(text)) return text.trim();
  if (text.includes('||')) {
    const parts = text.split('||').map((p) => p.trim()).filter(Boolean);
    if (parts.length >= 2) return `CONCAT(${parts.join(', ')})`;
  }
  if (text.includes('+') && /'[^']*'/.test(text)) {
    const parts = text.split('+').map((p) => p.trim()).filter(Boolean);
    if (parts.length >= 2) return `CONCAT(${parts.join(', ')})`;
  }
  return null;
}

function tryDirectSqlFunction(text: string, allowAggregates: boolean): string | null {
  const functions = allowAggregates ? [...WHITELISTED_FUNCTIONS, ...AGGREGATE_FUNCTIONS] : [...WHITELISTED_FUNCTIONS];
  const fnPattern = new RegExp(`\\b(${functions.join('|')})\\s*\\(`, 'i');
  if (fnPattern.test(text)) return text.trim();
  return null;
}

function tryArithmeticExpression(text: string): string | null {
  const trimmed = text.trim();
  if (/^[A-Za-z0-9_.\s+\-*/%()]+$/.test(trimmed) && /[+\-*/%]/.test(trimmed)) {
    return trimmed;
  }
  return null;
}

/**
 * Deterministically classifies a free-text transformation rule into a SQL-buildable
 * strategy, or MANUAL_REVIEW when it can't be safely turned into SQL. Order matters:
 * first matching strategy wins, then a token-whitelist gate can still downgrade to
 * MANUAL_REVIEW to avoid ever emitting a silently-wrong query.
 */
export function classifyTransformation(
  rawText: string,
  knownFields: string[],
  options: { allowAggregates?: boolean } = {}
): TransformationClassification {
  const allowAggregates = options.allowAggregates ?? false;
  const text = rawText.trim();

  if (!text || isTrivial(text)) {
    return { strategy: 'DIRECT_COPY', expression: null, rawText };
  }

  const attempts: Array<[TransformationStrategy, string | null]> = [
    ['CASE_EXPRESSION', tryCaseExpression(text)],
    ['DEFAULT_OR_LOOKUP', tryDefaultOrLookup(text)],
    ['CONCAT_EXPRESSION', tryConcatExpression(text)],
    ['DIRECT_SQL_FUNCTION', tryDirectSqlFunction(text, allowAggregates)],
    ['ARITHMETIC_EXPRESSION', tryArithmeticExpression(text)],
  ];

  for (const [strategy, expression] of attempts) {
    if (expression === null) continue;
    if (!isBalancedParens(expression)) continue;
    if (!containsOnlyWhitelistedTokens(expression, knownFields, allowAggregates)) continue;
    return { strategy, expression, rawText };
  }

  return { strategy: 'MANUAL_REVIEW', expression: null, rawText };
}

/** Replaces bare occurrences of known field names with `alias`.`field` qualified references. */
export function qualifyFieldReferences(expression: string, knownFields: string[], alias: string): string {
  let result = expression;
  const sortedFields = [...knownFields].sort((a, b) => b.length - a.length);
  for (const field of sortedFields) {
    if (!field) continue;
    const escaped = field.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    const pattern = new RegExp(`(?<!['"\`\\w])${escaped}(?!['"\`\\w])`, 'gi');
    result = result.replace(pattern, `${alias}.\`${field}\``);
  }
  return result;
}
